chirp-impl/src/main/java/sample/chirper/chirp/impl/JpaSessionImpl.java

```diff
@@ -0,0 +1,67 @@
+package sample.chirper.chirp.impl;
+
+import akka.Done;
+import com.lightbend.lagom.internal.persistence.jdbc.SlickProvider;
+import play.inject.ApplicationLifecycle;
+import scala.compat.java8.FutureConverters;
+import scala.compat.java8.JFunction0;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+import javax.persistence.Persistence;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+@Singleton
+public class JpaSessionImpl implements JpaSession {
+    private final EntityManagerFactory factory;
+    private final SlickProvider slick;
+
+    @Inject
+    JpaSessionImpl(SlickProvider slick, ApplicationLifecycle lifecycle) {
+        this.factory = Persistence.createEntityManagerFactory("default");
+        this.slick = slick;
+        lifecycle.addStopHook(this::close);
+    }
+
+    @Override
+    public <T> CompletionStage<T> withEntityManager(Function<EntityManager, T> block) {
+        return executeInSlickContext(() -> {
+            EntityManager entityManager = factory.createEntityManager();
+            try {
+                return block.apply(entityManager);
+            } finally {
+                entityManager.close();
+            }
+        });
+    }
+
+    @Override
+    public <T> CompletionStage<T> withTransaction(Function<EntityManager, T> block) {
+        return withEntityManager(entityManager -> {
+            EntityTransaction transaction = entityManager.getTransaction();
+            try {
+                transaction.begin();
+                T result = block.apply(entityManager);
+                transaction.commit();
+                return result;
+            } finally {
+                if (transaction.isActive()) transaction.rollback();
+            }
+        });
+    }
+
+    private CompletionStage<Done> close() {
+        return executeInSlickContext(() -> {
+            factory.close();
+            return Done.getInstance();
+        });
+    }
+
+    private <R> CompletionStage<R> executeInSlickContext(JFunction0<R> block) {
+        return FutureConverters.toJava(slick.db().io(block));
+    }
```

Comment:
(I destroyed my previous review by mistake). I think it's safe to run  `slick.db()` on the constructor and keep a reference `db` in a field.